modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/test/StartService.java

```diff
@@ -159,39 +164,162 @@ private void clearPendingConnectors() {
-                    Constants.ANN_NAME_CONFIG);
-            if (annotation != null) {
-                basePath = annotation.getAttributeValue(Constants.ANN_CONFIG_ATTR_BASE_PATH)
-                        .getStringValue();
-            }
-            if (basePath.startsWith("\"")) {
-                basePath = basePath.substring(1, basePath.length() - 1);
+            AnnAttachmentInfo annotationInfo = service.getAnnotationAttachmentInfo(Constants
+                    .HTTP_PACKAGE_PATH, Constants.ANN_NAME_CONFIG);
+
+            String basePath = discoverBasePathFrom(service, annotationInfo);
+            Set<ListenerConfiguration> listenerConfigurationSet = getDefaultOrDynamicListenerConfig(annotationInfo);
+
+            if (listenerConfigurationSet.isEmpty()) {
+                throw new BallerinaException(MSG_PREFIX + "Cannot find relevant transport listener configuration");
             }
 
-            if (!basePath.startsWith("/")) {
-                basePath = "/".concat(basePath);
+            //There can be multiple urls with the new transport implementation, so for the time being,
+            // just use the first listener configuration to generate the url
+            ListenerConfiguration config = listenerConfigurationSet.iterator().next();
+            String host = config.getHost().equals(DEFAULT_HOSTNAME) ? LOCALHOST : config.getHost();
+            int port = config.getPort();
+            String scheme = config.getScheme();
+            return new URL(scheme, host, port, basePath).toExternalForm();
+        } catch (MalformedURLException e) {
+            throw new BallerinaException(MSG_PREFIX + "Error while constructing service url for " + service.getName());
+        }
+
+    }
+
+    private Set<ListenerConfiguration> getDefaultOrDynamicListenerConfig(AnnAttachmentInfo annotationInfo) {
+        Map<String, Map<String, String>> listenerProp = buildListerProperties(annotationInfo);
+
+        Set<ListenerConfiguration> listenerConfigurationSet;
+        if (listenerProp == null || listenerProp.isEmpty()) {
+            listenerConfigurationSet =
+                    BallerinaConnectorManager.getInstance().getDefaultListenerConfiugrationSet();
+        } else {
+            listenerConfigurationSet = getListenerConfigurationsFrom(listenerProp);
+        }
+        return listenerConfigurationSet;
+    }
+
+    /**
+     * Method to build map of listener property maps given the service annotation attachment.
+     * This will first look for the port property and if present then it will get other properties,
+     * and create the property map.
+     *
+     * @param configInfo            In which listener configurations are specified.
+     * @return listenerConfMap      With required properties
+     */
+    private Map<String, Map<String, String>> buildListerProperties(AnnAttachmentInfo configInfo) {
+        if (configInfo == null) {
+            return null;
```

Comment:
Wouldn't it be better to return the empty map? that will help to avoid null pointers..